File: problem41.py

import math

num = 987654321
def primes_natural(n):
    li = []
    for j in range(2,n+1):
        flag =True
        for i in range(2,int(math.sqrt(j)) + 1):
            if(j%i == 0):
                flag = False
                break
        if(flag):
            li.append(j)
   

    return li





def prime_eratosthenes(n):
    prime_list = []
    primes = []
    for i in range(2, n+1):
        if i not in prime_list:
            
            primes.append(i)
            for j in range(i*i, n+1, i):
                prime_list.append(j)
    return primes

primes = primes_natural(num)
di = {
    1 : ['1'],
    2 : ['1','2'],
    3 : ['1','2','3'],
    4 : ['1','2','3','4'],
    5 : ['1','2','3','4','5'],
    6 : ['1','2','3','4','5','6'],
    7 : ['1','2','3','4','5','6','7'],
    8 : ['1','2','3','4','5','6','7','8'],
    9 : ['1','2','3','4','5','6','7','8','9'],
}

for i in primes:
    x = di[str(i).__len__()]
    y = list(set(str(i)))
    y.sort()
    if(x == y):
        print(i)

File: Problem41.cs
using System;
using System.Diagnostics;
using System.Collections.Generic;
using System.Collections;


namespace euler {
    class Problem41 {
        
        public static void Main(string[] args) {
            new Problem41().BruteForce();
            new Problem41().Digitsum();
        }

        public void BruteForce() {
            Stopwatch clock = Stopwatch.StartNew();

            int[] primes = ESieve(987654321);
            int result = 0;

            for (int i = primes.Length - 1; i >= 0; i--) {
                if(isPandigital(primes[i])){
                    result = primes[i];
                    break;
                }
            }
            
            clock.Stop();
            Console.WriteLine("The largest Pandigital prime is {0}", result);
            Console.WriteLine("Solution took {0} ms", clock.ElapsedMilliseconds);
        }

        public void Digitsum() {
            Stopwatch clock = Stopwatch.StartNew();

            int[] primes = ESieve(7654321);
            int result = 0;

            for (int i = primes.Length - 1; i >= 0; i--) {
                if (isPandigital(primes[i])) {
                    result = primes[i];
                    break;
                }
            }

            clock.Stop();
            Console.WriteLine("The largest Pandigital prime is {0}", result);
            Console.WriteLine("Solution took {0} ms", clock.ElapsedMilliseconds);
        }
     
        public int[] ESieve(int upperLimit) {

            int sieveBound = (int)(upperLimit - 1) / 2;
            int upperSqrt = ((int)Math.Sqrt(upperLimit) - 1) / 2;

            BitArray PrimeBits = new BitArray(sieveBound + 1, true);

            for (int i = 1; i <= upperSqrt; i++) {
                if (PrimeBits.Get(i)) {
                    for (int j = i * 2 * (i + 1); j <= sieveBound; j += 2 * i + 1) {
                        PrimeBits.Set(j, false);
                    }
                }
            }

            List<int> numbers = new List<int>((int)(upperLimit / (Math.Log(upperLimit) - 1.08366)));
            numbers.Add(2);
            for (int i = 1; i <= sieveBound; i++) {
                if (PrimeBits.Get(i)) {
                    numbers.Add(2 * i + 1);
                }
            }

            return numbers.ToArray();
        }

        private bool isPandigital(long n) {
            int digits = 0;
            int count = 0;
            int tmp;

            while (n > 0) {
                tmp = digits;
                digits = digits | 1 << (int)((n % 10) - 1); //The minus one is there to make 1 fill the first bit and so on
                if (tmp == digits) { //Check to see if the same digit is found multiple times
                    return false;
                }

                count++;
                n /= 10;
            }

            return digits == (1 << count) - 1;
        }

    }
}
